_posts/2024-09-05-releasing-apps.md

```diff
@@ -0,0 +1,84 @@
+
+### The one trunk to rule them all
+Managing two trunks was starting to get impractical - why not get rid of one of them?
+
+Trunk-based development is an approach where there’s one trunk branch (like `main`), and all work lands there. All work is releasable, made possible by feature flags and automated testing, meaning any given commit is safe to release! While this sounds great in theory, and works well in systems where rolling backwards or forwards is quick and easy, it’s problematic for mobile apps.
+
+Apps are constrained by the stores, not shipping our releases to customers before taking their time reviewing. It’s takes a good amount longer for adoption to then ramp up, meaning it takes a long time to create a fix and address it. Even if we had incredible test coverage, it’s near-impossible to test every given scenario your app will be run under in production.
+
+A branching structure growing in popularity for mobile apps nowadays is a combination of trunk-based development and release branches. Let’s explain:
+
+Developers do their work in small, manageable chunks, and merge into a trunk branch: `main`. Every week (or fortnight in our case), a workflow kicks off to cut a _release branch_. This release branch behaves similarly to the GitFlow release branch - it’s fixing our base at a given point, allowing us to stabilize before releasing. Fixes for the release could go one of two ways - based on release and merged into the release branch, or based on `main`, and back-ported to the release branch.
+
+// Insert branching diagram
+
+Merging a release branch back in to where it came from was one of the time sinks we were looking to get rid of, removing the need to resolve merge conflicts. Instead, we aimed for fixing on `main`, and _never_ merging the release branch back in.
+
+To make this “back-porting” simple for developers, we introduced an automatic cherry-picking workflow. When bugs come up in a release before deployment, pull requests with fixes would land in `main`, first. These pull requests have a special label which another piece of automation picks up on, knowing to cherry-pick the newly merged work on to the open release branch!
+
+This works the majority of the time, but what about merge conflicts? In our old school GitFlow model, resolving conflicts ended up being the responsibility of the release champion, who may or may not have all the context needed. With our cherry-picking flow, the conflicts end up having to be resolved more granularly the _other way around_.
+
+Developers fix their bugs on `main`, meaning the code they need to modify to address the issue has moved on from the base the release branch was cut from. If a fix can’t be cleanly cherry-picked onto the release branch, the author of the fix is prompted to create a pull request _per-fix_, addressing conflicts more granularly as they come up!
```

Comment:
Not sure if the branching diagram above will cover but might be nice to include this case as part of it! 

Or even two diagrams comparing the new with how it would've worked in the old approach where all bug fixes built up into the `release -> main` merge conflict time sink.

Reply:
Done in [d980c16](https://github.com/marksandspencer/marksandspencer.github.io/pull/18/commits/d980c165f6718127848a6cf1a18ce691fc6db1e7)